lib/awesome_print/formatters/ext/active_model_error_formatter.rb

```diff
@@ -0,0 +1,44 @@
+require_relative '../base_formatter'
+
+module AwesomePrint
+  module Formatters
+    class ActiveModelErrorFormatter < BaseFormatter
+
+      formatter_for :active_model_error
+
+      def self.formattable?(object)
+        defined?(::ActiveModel) && object.is_a?(::ActiveModel::Errors)
+      end
+
+      def format(object)
+        if @options[:raw]
+          return Formatters::ObjectFormatter.new(@inspector).format(object)
+        end
+
+        if !defined?(::ActiveSupport::OrderedHash)
+          return Formatters::SimpleFormatter.new(@inspector).format(object.inspect)
+        end
+
+        object_dump = object.marshal_dump.first
+
+        if object_dump.class.column_names != object_dump.attributes.keys
+          data = object_dump.attributes
+        else
+          data = object_dump.class.column_names.inject(::ActiveSupport::OrderedHash.new) do |hash, name|
+
+            if object_dump.has_attribute?(name) || object_dump.new_record?
+              value = object_dump.respond_to?(name) ? object_dump.send(name) : object_dump.read_attribute(name)
+              hash[name.to_sym] = value
+            end
+            hash
+          end
+        end
```

Comment:
Might consider breaking this out into a method, as it's getting a bit long. Perhaps something like this?

```ruby
def format(object)
  if @options[:raw]
    return Formatters::ObjectFormatter.new(@inspector).format(object)
  end

  if !defined?(::ActiveSupport::OrderedHash)
    return Formatters::SimpleFormatter.new(@inspector).format(object.inspect)
  end

  object_dump = object.marshal_dump.first

  data = if object_dump.class.column_names != object_dump.attributes.keys
    object_dump.attributes
  else
    extract_column_values(object_dump)
  end

  data.merge!(details: object.details, messages: object.messages)

  "#{object} " << Formatters::HashFormatter.new(@inspector).format(data)
end

def extract_column_values(object_dump)
  object_dump
    .class
    .column_names
    .each_with_object(::ActiveSupport::OrderedHash.new) do |column_name, hash|
      if object_dump.has_attribute?(name) || object_dump.new_record?
        value = object_dump.respond_to?(name) ? object_dump.send(name) : object_dump.read_attribute(name)
        hash[name.to_sym] = value
      end
    end
end
```

Reply:
I like @baweaver's method extraction and think I'd go a step further of extracting:

```ruby
data = if object_dump.class.column_names != object_dump.attributes.keys
    object_dump.attributes
  else
    extract_column_values(object_dump)
  end
```

to it's own method, returning `data`.